appModules/calculatorapp: add support for Windows 11 Calculator

josephsl opened the pull request:
### Link to issue number:
Closes #13212 

### Summary of the issue:
Adds support for Windows 11 Calculator (calculatorapp.exe), consisting of an alias app module and responding to UI changes via Windows 10 Calculator app module.

### Description of how this pull request fixes the issue:
In Windows 11, Calculator not only saw UI changes, but it comes with a renamed executable (calculator.exe -> calculatorapp.exe). To take care of these changes:

1. Windows 11 Calculator app module was created as an alias for Windows 10 Calculator to follow precedence set by other apps such as emoji panel, Windows Search, and Mail.
2. Windows 10 Calculator, the base app module, provides detection of Windows 11 UI (if Windows 11 version is in use, UIA notification event handler's results detection code must descend one more level to locate display element).

An alternative is a dedicated Windows 11 Calculator app module that inherits from Windows 10 Calculator's AppModule class and tweaks UIA notification event handler. Although this makes it easy to identify which app module is which, it introduces code duplication which might be an issue from maintenance point of view. Therefore, an alias app module and a tweaked base app module is more suitable for this situation.

### Testing strategy:
Manual testing: requires Windows 11, make sure NVDA performs as expected while using Calculator app just as in Windows 10 (namely not nanouncing display text unless Enter/Escape/Delete is pressed, unit converter results are announced, and others).

### Known issues with pull request:
None

### Change log entries:
New features:

Added support for Windows 11 Calculator. (#13212)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

josephsl committed:
appModules/calculator: descend one more level when handling UIA notification event in Windows 11 version. Re #13212.

In Windows 11 Calculator, due to UI redesign, app interface lives inside the last child of the foreground window (foregroundObject.children[1].lastChild.firstChild), therefore check product version and descend one more level if this is Windows 11 version. This is only one of several changes introduced in Windows 11 Calculator.

josephsl committed:
appModules/calculatorapp: add Windows 11 Calculator app module as an alias for Windows 10 Calculator. Re #13212.

The biggest change between Windows 10 and 11 Calculator app is renamed executabel - calculator.exe -> calculatorapp.exe. Following precedence from other Windows apps, add Windows 11 Calculator as an alias for Windows 10 version, with the latter also adding version checks for Windows 11 version (similar in arrangemenet to emoji panel support in Windows 10/11).

feerrenrut commented on source/appModules/calculator.py:
```diff
@@ -76,6 +76,10 @@ def event_UIA_notification(self, obj, nextHandler, displayString=None, activityI
 			)
 		except AttributeError:
 			resultElement = api.getForegroundObject().children[1].lastChild
+			# Descend one more time in Windows 11 Calculator.
+			calculatorVersion = int(self.productVersion.split(".")[0])
+			if calculatorVersion >= 11:
+				resultElement = resultElement.firstChild
```
Why is this all in an exception handler?

josephsl commented:
Hi, it is a legacy of old releases where UIA notification event can come from multiple elements, and in some cases, elements are next to unit converter interface. After testing the event handler, it turns out attribute error check is not needed in recent releases, therefore I’m going to rework this and apply a point about assigning product version somewhere (likely at the start of the event handler). Thanks for reminding me to stay updated on things.

josephsl committed:
appModules/calculator: restructure UIA notification event handler to remove legacy try/except statements.

Reviewed by Reef Turner (NV Access): A try/AttributeError exception check was present in order to support older Windows 10 Calculator releases, mostly to detect notifications coming from unit converters. As of January 2022, this legacy code is no longer needed as legacy code is used in Calculator releases in older Windows 10 feature updates such as Version 1709 which is unsupported (the latest version of Windows 10 is 10.2103). Besides, version variable was defined in attribute error exception handler, which isn't ideal and makes the code harder to read. Therefore remove the redundant try/except statements and fetch results screen as part of 'DisplayUpdated' activity Id handling (after all, results screen is fetched when this activity Id is used).

josephsl committed:
appModules/calculator: move product version str to int conversion to top of UIA notification event handler.

josephsl commented:
@seanbudd, ping, milestone update please i.e. when do you expect this will be merged into master. Thanks.

feerrenrut commented on source/appModules/calculator.py:
```diff
@@ -66,28 +66,25 @@ def event_nameChange(self, obj, nextHandler):
+		# #12268: for "DisplayUpdated", announce display strings in braille  no matter what they are.
 		if activityId == "DisplayUpdated":
 			braille.handler.message(displayString)
-		try:
-			shouldAnnounceNotification = (
-				obj.previous.UIAAutomationId in
-				("numberPad", "UnitConverterRootGrid")
-			)
-		except AttributeError:
```
Removing this effectively changes the scope of the subsequent code, is the indent level as expected?

josephsl commented:
Hi, yes, the indent level should be kept as we are effectively dealing with a single activity Id. Thanks.

feerrenrut commented on source/appModules/calculator.py:
```diff
@@ -66,28 +66,25 @@ def event_nameChange(self, obj, nextHandler):
 
 	def event_UIA_notification(self, obj, nextHandler, displayString=None, activityId=None, **kwargs):
-		# #12268: for "DisplayUpdated", announce display strings in braille and move on.
+		calculatorVersion = int(self.productVersion.split(".")[0])
+		# #12268: for "DisplayUpdated", announce display strings in braille  no matter what they are.
```
In that case, @josephsl could you add another comment below this one to say that the only type of activity that should be handled here is "DisplayUpdated". Perhaps an example of other activities that shouldn't be handled, or an explanation about why other activities are not relevant.

josephsl commented:
Hi, will do so in the next commit. Thanks.

josephsl committed:
Merge branch 'master' into win11calculator

josephsl committed:
apModules/calculator: add additional comment on activity Id's.

Reviewed by Reef Turner (NV Access): 'DisplayUpdated' activity Id is responsible for displaying calculator content and is updated when numbers are entered (pressing Enter and Delete key does update display content, but for purposes of handling UIA notification event, deal with number entry). Therefore add a comment that notes other activity Id's that are present but not handled in the event handler.

feerrenrut reviewed (approved):
Thanks @josephsl

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into win11calculator

feerrenrut committed:
update changes file for PR #13215
